README.md
# Logstash Plugin

This is a plugin for [Logstash](https://github.com/elastic/logstash).

It is fully free and fully open source. The license is Apache 2.0, meaning you are pretty much free to use it however you want in whatever way.

## Documentation

Logstash provides infrastructure to automatically generate documentation for this plugin. We use the asciidoc format to write documentation so any comments in the source code will be first converted into asciidoc and then into html. All plugin documentation are placed under one [central location](http://www.elastic.co/guide/en/logstash/current/).

- For formatting code or config example, you can use the asciidoc `[source,ruby]` directive
- For more asciidoc formatting tips, see the excellent reference here https://github.com/elastic/docs#asciidoc-guide

## Need Help?

Need help? Try #logstash on freenode IRC or the https://discuss.elastic.co/c/logstash discussion forum.

## Developing

### 1. Plugin Development and Testing

#### Code
- To get started, you'll need JRuby with the Bundler gem installed.

- Create a new plugin or clone and existing from the GitHub [logstash-plugins](https://github.com/logstash-plugins) organization. We also provide [example plugins](https://github.com/logstash-plugins?query=example).

- Install dependencies
```sh
bundle install
```

#### Test

- Update your dependencies

```sh
bundle install
```

- Run tests

```sh
bundle exec rspec
```

### 2. Running your unpublished Plugin in Logstash

#### 2.1 Run in a local Logstash clone

- Edit Logstash `Gemfile` and add the local plugin path, for example:
```ruby
gem "logstash-filter-awesome", :path => "/your/local/logstash-filter-awesome"
```
- Install plugin
```sh
bin/logstash-plugin install --no-verify
```
- Run Logstash with your plugin
```sh
bin/logstash -e 'filter {awesome {}}'
```
At this point any modifications to the plugin code will be applied to this local Logstash setup. After modifying the plugin, simply rerun Logstash.

#### 2.2 Run in an installed Logstash

You can use the same **2.1** method to run your plugin in an installed Logstash by editing its `Gemfile` and pointing the `:path` to your local plugin development directory or you can build the gem and install it using:

- Build your plugin gem
```sh
gem build logstash-filter-awesome.gemspec
```
- Install the plugin from the Logstash home
```sh
bin/logstash-plugin install /your/local/plugin/logstash-filter-awesome.gem
```
- Start Logstash and proceed to test the plugin

## Contributing

All contributions are welcome: ideas, patches, documentation, bug reports, complaints, and even something you drew up on a napkin.

Programming is not a required skill. Whatever you've seen about open source and maintainers or community members  saying "send patches or die" - you will not see that here.

It is more important to the community that you are able to contribute.

For more information about contributing, see the [CONTRIBUTING](https://github.com/elastic/logstash/blob/master/CONTRIBUTING.md) file.


CHANGELOG.md
## 1.3.3
  - Silence warnings when loading dictionary MIB files [#118](https://github.com/logstash-plugins/logstash-input-snmp/pull/118)

## 1.3.2
  -  [DOC] Add troubleshooting help for "failed to locate MIB module" error when using smidump to convert MIBs [#112](https://github.com/logstash-plugins/logstash-input-snmp/pull/112)
  -  Fix CI failures [#113](https://github.com/logstash-plugins/logstash-input-snmp/pull/113)

## 1.3.1
  -  Refactor: handle no response(s) wout error logging [#105](https://github.com/logstash-plugins/logstash-input-snmp/pull/105)

## 1.3.0
  - Feat: ECS compliance + optional target [#99](https://github.com/logstash-plugins/logstash-input-snmp/pull/99)
  - Internal: update to Gradle 7 [#102](https://github.com/logstash-plugins/logstash-input-snmp/pull/102)

## 1.2.8
  - Fixed interval handling to only sleep off the _remainder_ of the interval (if any), and to log a helpful warning when crawling the hosts takes longer than the configured interval [#100](https://github.com/logstash-plugins/logstash-input-snmp/pull/100). Fixes [#61](https://github.com/logstash-plugins/logstash-input-snmp/issues/61).

## 1.2.7
  - Added integration tests to ensure SNMP server and IPv6 connections [#90](https://github.com/logstash-plugins/logstash-input-snmp/issues/90). Fixes[#87](https://github.com/logstash-plugins/logstash-input-snmp/issues/87).

## 1.2.6
  - [DOC] Add example on setting IPv6 hosts [#89](https://github.com/logstash-plugins/logstash-input-snmp/pull/89)

## 1.2.5
  - Updated snmp4j library to v2.8.4 [#86](https://github.com/logstash-plugins/logstash-input-snmp/pull/86)

## 1.2.4
  - Fixed: support SNMPv3 multiple identical security name with different credentials [#84](https://github.com/logstash-plugins/logstash-input-snmp/pull/84)

## 1.2.3
  - Fixed: multithreading problem when using multiple snmp inputs with multiple v3 credentials [#80](https://github.com/logstash-plugins/logstash-input-snmp/pull/80)

## 1.2.2
  - Refactor: scope and review java_imports [#72](https://github.com/logstash-plugins/logstash-input-snmp/pull/72)

## 1.2.1
  - Fixed GAUGE32 integer overflow [#65](https://github.com/logstash-plugins/logstash-input-snmp/pull/65)

## 1.2.0
  - Adding oid_path_length config option [#59](https://github.com/logstash-plugins/logstash-input-snmp/pull/59)
  - Fixing bug with table support removing index value from OIDs [#60])https://github.com/logstash-plugins/logstash-input-snmp/issues/60)

## 1.1.1
  - Added information and other improvements to documentation [#57](https://github.com/logstash-plugins/logstash-input-snmp/pull/57)

## 1.1.0
  - Added support for querying SNMP tables [#49](https://github.com/logstash-plugins/logstash-input-snmp/pull/49)
  - Changed three error messages in the base_client to include the target address for clarity in the logs.

## 1.0.1
  - Added no_codec condition to the documentation and bumped version [#39](https://github.com/logstash-plugins/logstash-input-snmp/pull/39)
  - Changed docs to improve options layout [#38](https://github.com/logstash-plugins/logstash-input-snmp/pull/38)

## 1.0.0
  - Added improved syntax coercion [#32](https://github.com/logstash-plugins/logstash-input-snmp/pull/32)

## 0.1.0.beta5
  - Added OPAQUE type coercion [#29](https://github.com/logstash-plugins/logstash-input-snmp/pull/29)
  - Added SNMPv3 support [#27](https://github.com/logstash-plugins/logstash-input-snmp/pull/27)
  - Added support for provided MIBs [#25](https://github.com/logstash-plugins/logstash-input-snmp/pull/25)

## 0.1.0.beta4
  - Fixed missing coercions [#12](https://github.com/logstash-plugins/logstash-input-snmp/pull/12)

## 0.1.0.beta3
  - add tcp transport protocol support, https://github.com/logstash-plugins/logstash-input-snmp/pull/8
  - add SNMPv1 protocol version support, https://github.com/logstash-plugins/logstash-input-snmp/pull/9

## 0.1.0.beta2
  - add host info in metadata and host field, https://github.com/logstash-plugins/logstash-input-snmp/pull/7

## 0.1.0.beta1
  - First beta version


test/loggable.rb
# encoding: utf-8
require "logstash/logging/logger"
require "logstash/namespace"

module LogStash module Util
  module Loggable
    def self.included(klass)

      def klass.log4j_name
        ruby_name = self.name || self.class.name || self.class.to_s
        ruby_name.gsub('::', '.').downcase
      end

      def klass.logger
        @logger ||= LogStash::Logging::Logger.new(log4j_name)
      end

      def klass.slow_logger(warn_threshold, info_threshold, debug_threshold, trace_threshold)
        @slow_logger ||= LogStash::Logging::SlowLogger.new(log4j_name, warn_threshold, info_threshold, debug_threshold, trace_threshold)
      end

      def logger
        self.class.logger
      end

      def slow_logger(warn_threshold, info_threshold, debug_threshold, trace_threshold)
        self.class.slow_logger(warn_threshold, info_threshold, debug_threshold, trace_threshold)
      end
    end
  end
end; end

test/test_mib.rb
# encoding: utf-8
$LOAD_PATH.unshift(File.expand_path(File.join(__FILE__, "..", "..", "lib")))

require_relative "loggable"
require "pp"
require "logstash/inputs/snmp/mib"

mib = LogStash::SnmpMib.new
# module_name, names, oids = mib.read_mib_dic("tmp/RFC1213-MIB.dic")
# puts(module_name)
# pp(names)
# pp(oids)

mib.add_mib_path(File.expand_path(File.join("..", "..", "spec", "fixtures", "RFC1213-MIB.dic"), __FILE__))

pp mib.map_oid("1.3.6.1.2.1.1")
pp mib.map_oid("1.3.6.1.2.1.1.1.0")

test/test_client.rb
# encoding: utf-8
$LOAD_PATH.unshift(File.expand_path(File.join(__FILE__, "..", "..", "lib")))

require_relative "loggable"
require "pp"
require "logstash/inputs/snmp/client"
require "logstash/inputs/snmp/clientv3"
require "logstash/inputs/snmp/mib"

mib = LogStash::SnmpMib.new
mib.add_mib_path(File.expand_path(File.join("..", "..", "spec", "fixtures", "RFC1213-MIB.dic"), __FILE__))

#client = LogStash::SnmpClient.new("tcp", "127.0.0.1", "1161", "public", "2c", 2, 1000, mib)

puts("-- SNMPv2c")
client = LogStash::SnmpClient.new("udp", "127.0.0.1", "161", "public", "2c", 2, 1000, mib)

pp client.get("1.3.6.1.2.1.1.1.0")
pp client.get("1.3.6.1.2.1.1.3.0")
pp client.get("1.3.6.1.2.1.1.5.0")
pp client.walk("1.3.6.1.2.1.1")


puts("-- SNMPv3")
client = LogStash::SnmpClientV3.new("udp", "127.0.0.1", "161", 2, 10000, mib, "rouser", "SHA", "abcd1234", "AES", "efgh5678", "authpriv")

pp client.get("1.3.6.1.2.1.1.1.0")
pp client.get("1.3.6.1.2.1.1.3.0")
pp client.get("1.3.6.1.2.1.1.5.0")
pp client.walk("1.3.6.1.2.1.1")


lib/logstash/inputs/snmp/mib.rb
# encoding: utf-8

require "logstash/util/loggable"

module LogStash
  class SnmpMibError < StandardError
  end

  class SnmpMib
    include LogStash::Util::Loggable

    attr_reader :tree

    class Oid
      def self.parse(oid)
        oid.split(".").map(&:to_i)
      end
    end

    class BaseNode
      attr_accessor :name, :childs

      def initialize(name)
        @name = name
        @childs = []
      end
    end

    class Node < BaseNode
      attr_reader :node_type, :module_name, :oid, :oid_path

      def initialize(node_type, name, module_name, oid)
        super(name)
        @node_type = node_type
        @module_name = module_name
        @oid = oid
        @oid_path = Oid.parse(oid)
      end
    end

    class Tree
      include LogStash::Util::Loggable

      def initialize
        @root = BaseNode.new("root")
      end

      def add_node(node)
        current = @root
        path = node.oid_path.dup

        # follow the OID path up until but not including the last node
        # and add intermediate missing nodes if needed
        last_node = path.pop
        path.each do |i|
          if current.childs[i].nil?
            current.childs[i] = BaseNode.new(i.to_s)
          end
          current = current.childs[i]
        end

        if current.childs[last_node] && current.childs[last_node].name != node.name
          logger.debug("warning: overwriting MIB OID '#{node.oid}' and name '#{current.childs[last_node].name}' with new name '#{node.name}' from module '#{node.module_name}'")
         end

        # preserve previous childs if replacing a node
        node.childs = current.childs[last_node].childs.dup if current.childs[last_node]

        current.childs[last_node] = node
      end

      def map_oid(oid, strip_root = 0, path_length = 0)
        path = Oid.parse(oid)

        result = []
        node = @root

        loop do
          break if path.empty?
          i = path.shift

          node = node.childs[i]

          if node.nil?
            result += path.unshift(i)
            break
          end
          result << node.name
        end

	if strip_root > 0
	  result.drop(strip_root).join(".")
	elsif path_length > 0
	  result.pop(path_length).join(".")
	else
	  result.join(".")
	end
      end
    end

    def initialize
      @tree = Tree.new
    end

    # add a specific mib dic file or all mib dic files of the given directory to the current mib database
    # @param path [String] a file or directory path to mib dic file(s)
    # @return [Array] array of warning strings if any OID or name has been overwritten or the empty array when no warning
    def add_mib_path(path)
      dic_files = if ::File.directory?(path)
        Dir[::File.join(path, "*.dic")]
      elsif ::File.file?(path)
        [path]
      else
        raise(SnmpMibError, "file or directory path expected: #{path.to_s}")
      end

      dic_files.each do |f|
        module_name, nodes = read_mib_dic(f)
        nodes.each do |k, v|
          @tree.add_node(Node.new(v["nodetype"], k, v["moduleName"], v["oid"]))
        end
      end
    end

    # read and parse a mib dic file
    #
    # @param filename [String] file path of a mib dic file
    # @return [[String, Hash, Hash, Hash]] the 2-tuple of the mib module name and the complete nodes
    def read_mib_dic(filename)
      mib = eval_mib_dic(filename)

      raise(SnmpMibError, "invalid mib dic format for file #{filename}") unless mib
      module_name = mib["moduleName"]

      raise(SnmpMibError, "invalid mib dic format for file #{filename}") unless module_name
      nodes = mib["nodes"] || []

      if nodes.empty?
        logger.warn("no nodes defined in mib dic file #{filename}")
      end

      [module_name, nodes]
    end

    def map_oid(oid, strip_root = 0, path_length = 0)
      @tree.map_oid(oid, strip_root, path_length)
    end

    private

    def eval_mib_dic(filename)
      mib_dic = IO.read(filename)
      mib_hash = mib_dic.
        gsub(':', '=>').                  # fix hash syntax
        gsub('(', '[').gsub(')', ']').    # fix tuple syntax
        sub('FILENAME =', 'filename =').  # get rid of constants
        sub('MIB =', 'mib =')

      mib = nil
      silence_warnings do
        eval(mib_hash)
      end
      mib
    rescue Exception => e
      # rescuing Exception class is important here to rescue SyntaxError from eval
      raise(SnmpMibError, "error parsing mib dic file: #{filename}, error: #{e.message}")
    end

    # MIB definition files may have duplicate entries that, since Ruby 2.7, generate warnings when they are converted to a hash.
    # This method will suppress these warnings, unless `debug` logging is enabled.
    def silence_warnings
      warn_level = $VERBOSE
      $VERBOSE = logger.debug? ? warn_level : nil
      yield
    ensure
      $VERBOSE = warn_level
    end
  end
end


lib/logstash/inputs/snmp/client.rb
require "java"
require "logstash-input-snmp_jars.rb"
require_relative "base_client"

module LogStash
  class SnmpClient < BaseSnmpClient

    java_import "org.snmp4j.CommunityTarget"
    java_import "org.snmp4j.PDU"
    java_import "org.snmp4j.ScopedPDU"
    java_import "org.snmp4j.Snmp"
    java_import "org.snmp4j.Target"
    java_import "org.snmp4j.smi.Address"
    java_import "org.snmp4j.smi.GenericAddress"
    java_import "org.snmp4j.smi.OctetString"
    java_import "org.snmp4j.util.DefaultPDUFactory"

    def initialize(protocol, address, port, community, version, retries, timeout, mib)
      super(protocol, address, port, retries, timeout, mib)
      raise(SnmpClientError, "SnmpClient is expecting version '1' or '2c'") unless ["1", "2c"].include?(version.to_s)

      @snmp = Snmp.new(create_transport(protocol))
      @snmp.listen

      @target = build_target("#{protocol}:#{address}/#{port}", community, version, retries, timeout)
    end

    def close
      @snmp.close
    end

    private

    def get_pdu
      pdu = PDU.new
      pdu.setType(PDU::GET)
      pdu
    end

    def get_pdu_factory
      DefaultPDUFactory.new
    end

    def build_target(address, community, version, retries, timeout)
      target = CommunityTarget.new
      target.setCommunity(OctetString.new(community))
      target.setAddress(GenericAddress.parse(address))
      target.setRetries(retries)
      target.setTimeout(timeout)
      target.setVersion(parse_version(version))
      target
    end
  end
end


lib/logstash/inputs/snmp.rb
# encoding: utf-8
require "logstash/inputs/base"
require "logstash/namespace"
require "stud/interval"
require "socket" # for Socket.gethostname
require_relative "snmp/client"
require_relative "snmp/clientv3"
require_relative "snmp/mib"

require 'logstash/plugin_mixins/ecs_compatibility_support'
require 'logstash/plugin_mixins/ecs_compatibility_support/target_check'
require 'logstash/plugin_mixins/event_support/event_factory_adapter'
require 'logstash/plugin_mixins/validator_support/field_reference_validation_adapter'

# Generate a repeating message.
#
# This plugin is intented only as an example.

class LogStash::Inputs::Snmp < LogStash::Inputs::Base

  include LogStash::PluginMixins::ECSCompatibilitySupport(:disabled, :v1, :v8 => :v1)
  include LogStash::PluginMixins::ECSCompatibilitySupport::TargetCheck

  include LogStash::PluginMixins::EventSupport::EventFactoryAdapter

  extend LogStash::PluginMixins::ValidatorSupport::FieldReferenceValidationAdapter

  config_name "snmp"

  # List of OIDs for which we want to retrieve the scalar value
  config :get,:validate => :array # ["1.3.6.1.2.1.1.1.0"]

  # List of OIDs for which we want to retrieve the subtree of information
  config :walk,:validate => :array # ["1.3.6.1.2.1.1.1.0"]

  # List of tables to walk
  config :tables, :validate => :array  #[ {"name" => "interfaces" "columns" => ["1.3.6.1.2.1.2.2.1.1", "1.3.6.1.2.1.2.2.1.2", "1.3.6.1.2.1.2.2.1.5"]} ]

  # List of hosts to query the configured `get` and `walk` options.
  #
  # Each host definition is a hash and must define the `host` key and value.
  #  `host` must use the format {tcp|udp}:{ip address}/{port}
  #  for example `host => "udp:127.0.0.1/161"`
  # Each host definition can optionally include the following keys and values:
  #  `community` with a default value of `public`
  #  `version` `1`, `2c` or `3` with a default value of `2c`
  #  `retries` with a default value of `2`
  #  `timeout` in milliseconds with a default value of `1000`
  config :hosts, :validate => :array  #[ {"host" => "udp:127.0.0.1/161", "community" => "public"} ]

  # This plugin provides sets of MIBs publicly available. The full paths to these provided MIBs paths
  # Will be displayed at plugin startup.
  config :use_provided_mibs, :validate => :boolean, :default => true

  # List of paths of MIB .dic files of dirs. If a dir path is specified, all files with .dic extension will be loaded.
  #
  # ATTENTION: a MIB .dic file must be generated using the libsmi library `smidump` command line utility
  # like this for example. Here the `RFC1213-MIB.txt` file is an ASN.1 MIB file.
  #
  # `$ smidump -k -f python RFC1213-MIB.txt > RFC1213-MIB.dic`
  #
  # The OSS libsmi library https://www.ibr.cs.tu-bs.de/projects/libsmi/ is available & installable
  # on most OS.
  config :mib_paths, :validate => :array # ["path/to/mib.dic", "path/to/mib/dir"]

  # number of OID root digits to ignore in event field name. For example, in a numeric OID
  # like 1.3.6.1.2.1.1.1.0" the first 5 digits could be ignored by setting oid_root_skip => 5
  # which would result in a field name "1.1.1.0". Similarly when a MIB is used an OID such
  # as "1.3.6.1.2.mib-2.system.sysDescr.0" would become "mib-2.system.sysDescr.0"
  config :oid_root_skip, :validate => :number, :default => 0

  # number of OID tail digits to retain in event field name. For example, in a numeric OID
  # like 1.3.6.1.2.1.1.1.0" the last 2 digits could be retained by setting oid_path_length => 2
  # which would result in a field name "1.0". Similarly, when a MIB is used an OID such as
  # "1.3.6.1.2.mib-2.system.sysDescr.0" would become "sysDescr.0"
  config :oid_path_length, :validate => :number, :default => 0

  # Set polling interval in seconds
  #
  # The default, `30`, means poll each host every 30 seconds.
  config :interval, :validate => :number, :default => 30

  # SNMPv3 Credentials
  #
  # A single user can be configured and will be used for all defined SNMPv3 hosts.
  # Multiple snmp input declarations will be needed if multiple SNMPv3 users are required.
  # If not using SNMPv3 simply leave options empty.

  # The SNMPv3 security name or user name
  config :security_name, :validate => :string

  # The SNMPv3 authentication protocol or type
  config :auth_protocol, :validate => ["md5", "sha", "sha2", "hmac128sha224", "hmac192sha256", "hmac256sha384", "hmac384sha512"]

  # The SNMPv3 authentication passphrase or password
  config :auth_pass, :validate => :password

  # The SNMPv3 privacy/encryption protocol
  config :priv_protocol, :validate => ["des", "3des", "aes", "aes128", "aes192", "aes256"]

  # The SNMPv3 encryption password
  config :priv_pass, :validate => :password

  # The SNMPv3 security level can be Authentication, No Privacy; Authentication, Privacy; or no Authentication, no Privacy
  config :security_level, :validate => ["noAuthNoPriv", "authNoPriv", "authPriv"]

  # Defines a target field for placing fields.
  # If this setting is omitted, data gets stored at the root (top level) of the event.
  # The target is only relevant while decoding data into a new event.
  config :target, :validate => :field_reference

  BASE_MIB_PATH = ::File.join(__FILE__, "..", "..", "..", "mibs")
  PROVIDED_MIB_PATHS = [::File.join(BASE_MIB_PATH, "logstash"), ::File.join(BASE_MIB_PATH, "ietf")].map { |path| ::File.expand_path(path) }

  def initialize(params={})
    super(params)

    @host_protocol_field = ecs_select[disabled: '[@metadata][host_protocol]', v1: '[@metadata][input][snmp][host][protocol]']
    @host_address_field = ecs_select[disabled: '[@metadata][host_address]', v1: '[@metadata][input][snmp][host][address]']
    @host_port_field = ecs_select[disabled: '[@metadata][host_port]', v1: '[@metadata][input][snmp][host][port]']
    @host_community_field = ecs_select[disabled: '[@metadata][host_community]', v1: '[@metadata][input][snmp][host][community]']

    # Add the default "host" field to the event, for backwards compatibility, or host.ip in ecs mode
    unless params.key?('add_field')
      host_ip_field = ecs_select[disabled: "host", v1: "[host][ip]"]
      @add_field = { host_ip_field => "%{#{@host_address_field}}" }
    end
  end

  def register
    validate_oids!
    validate_hosts!
    validate_tables!
    validate_strip!

    mib = LogStash::SnmpMib.new

    if @use_provided_mibs
      PROVIDED_MIB_PATHS.each do |path|
        logger.info("using plugin provided MIB path #{path}")
        mib.add_mib_path(path)
      end
    end

    Array(@mib_paths).each do |path|
      logger.info("using user provided MIB path #{path}")
      mib.add_mib_path(path)
    end

    # setup client definitions per provided host

    @client_definitions = []
    @hosts.each do |host|
      host_name = host["host"]
      community = host["community"] || "public"
      version = host["version"] || "2c"
      raise(LogStash::ConfigurationError, "only protocol version '1', '2c' and '3' are supported for host option '#{host_name}'") unless version =~ VERSION_REGEX

      retries = host["retries"]&.to_i || 2
      timeout = host["timeout"]&.to_i || 1000

      # TODO: move these validations in a custom validator so it happens before the register method is called.
      host_details = host_name.match(HOST_REGEX)
      raise(LogStash::ConfigurationError, "invalid format for host option '#{host_name}'") unless host_details
      raise(LogStash::ConfigurationError, "only udp & tcp protocols are supported for host option '#{host_name}'") unless host_details[:host_protocol].to_s =~ /^(?:udp|tcp)$/i

      protocol = host_details[:host_protocol]
      address = host_details[:host_address]
      port = host_details[:host_port]

      definition = {
        :get => Array(get),
        :walk => Array(walk),

        :host_protocol => protocol,
        :host_address => address,
        :host_port => port,
        :host_community => community,
      }

      if version == "3"
        validate_v3_user! # don't really care if verified for every host
        auth_pass = @auth_pass.nil? ? nil : @auth_pass.value
        priv_pass = @priv_pass.nil? ? nil : @priv_pass.value
        definition[:client] = LogStash::SnmpClientV3.new(protocol, address, port, retries, timeout, mib, @security_name, @auth_protocol, auth_pass, @priv_protocol, priv_pass, @security_level)
      else
        definition[:client] = LogStash::SnmpClient.new(protocol, address, port, community, version, retries, timeout, mib)
      end
      @client_definitions << definition
    end
  end

  def run(queue)
    # for now a naive single threaded poller which sleeps off the remaining interval between
    # each run. each run polls all the defined hosts for the get, table and walk options.
    stoppable_interval_runner.every(@interval, "polling hosts") do
      poll_clients(queue)
    end
  end

  def poll_clients(queue)
    @client_definitions.each do |definition|
      client = definition[:client]
      host = definition[:host_address]
      result = {}

      if !definition[:get].empty?
        oids = definition[:get]
        begin
          data = client.get(oids, @oid_root_skip, @oid_path_length)
          if data
            result.update(data)
          else
            logger.debug? && logger.debug("get operation returned no response", host: host, oids: oids)
          end
        rescue => e
          logger.error("error invoking get operation, ignoring", host: host, oids: oids, exception: e, backtrace: e.backtrace)
        end
      end

      if !definition[:walk].empty?
        definition[:walk].each do |oid|
          begin
            data = client.walk(oid, @oid_root_skip, @oid_path_length)
            if data
              result.update(data)
            else
              logger.debug? && logger.debug("walk operation returned no response", host: host, oid: oid)
            end
          rescue => e
            logger.error("error invoking walk operation, ignoring", host: host, oid: oid, exception: e, backtrace: e.backtrace)
          end
        end
      end

      if !Array(@tables).empty?
        @tables.each do |table_entry|
          begin
            data = client.table(table_entry, @oid_root_skip, @oid_path_length)
            if data
              result.update(data)
            else
              logger.debug? && logger.debug("table operation returned no response", host: host, table: table_entry)
            end
          rescue => e
            logger.error("error invoking table operation, ignoring",
                         host: host, table_name: table_entry['name'], exception: e, backtrace: e.backtrace)
          end
        end
      end

      unless result.empty?
        event = targeted_event_factory.new_event(result)
        event.set(@host_protocol_field, definition[:host_protocol])
        event.set(@host_address_field, definition[:host_address])
        event.set(@host_port_field, definition[:host_port])
        event.set(@host_community_field, definition[:host_community])
        decorate(event)
        queue << event
      else
        logger.debug? && logger.debug("no snmp data retrieved", host: definition[:host_address])
      end
    end
  end

  def stoppable_interval_runner
    StoppableIntervalRunner.new(self)
  end

  def close
    @client_definitions.each do |definition|
      begin
        definition[:client].close
      rescue => e
        logger.warn("error closing client on #{definition[:host_address]}, ignoring", :exception => e)
      end
    end
  end

  def stop
  end

  private

  OID_REGEX = /^\.?([0-9\.]+)$/
  HOST_REGEX = /^(?<host_protocol>\w+):(?<host_address>.+)\/(?<host_port>\d+)$/i
  VERSION_REGEX =/^1|2c|3$/

  def validate_oids!
    @get = Array(@get).map do |oid|
      # verify oids for valid pattern and get rid or any leading dot if present
      unless oid =~ OID_REGEX
        raise(LogStash::ConfigurationError, "The get option oid '#{oid}' has an invalid format")
      end
      $1
    end

    @walk = Array(@walk).map do |oid|
      # verify oids for valid pattern and get rid or any leading dot if present
      unless oid =~ OID_REGEX
        raise(LogStash::ConfigurationError, "The walk option oid '#{oid}' has an invalid format")
      end
      $1
    end

    if !@tables.nil?
      @tables.each do |table_entry|
      # Verify oids for valid pattern and get rid of any leading dot if present
        columns = table_entry["columns"]
        columns.each do |column|
          unless column =~ OID_REGEX
      	    raise(Logstash::ConfigurationError, "The table column oid '#{column}' is an invalid format")
          end
        end
        $1
      end
    end

    raise(LogStash::ConfigurationError, "at least one get OID, one walk OID, or one table OID is required") if @get.empty? && @walk.empty? && @tables.nil?
  end

  def validate_v3_user!
    errors = []

    errors << "v3 user must have a \"security_name\" option" if @security_name.nil?
    errors << "you must specify an auth protocol if you specify an auth pass" if @auth_protocol.nil? && !@auth_pass.nil?
    errors << "you must specify an auth pass if you specify an auth protocol" if !@auth_protocol.nil? && @auth_pass.nil?
    errors << "you must specify a priv protocol if you specify a priv pass" if @priv_protocol.nil? && !@priv_pass.nil?
    errors << "you must specify a priv pass if you specify a priv protocol" if !@priv_protocol.nil? &&  @priv_pass.nil?

    raise(LogStash::ConfigurationError, errors.join(", ")) unless errors.empty?
   end


  def validate_hosts!
    # TODO: for new we only validate the host part, not the other optional options

    raise(LogStash::ConfigurationError, "at least one host definition is required") if Array(@hosts).empty?

    @hosts.each do |host|
      raise(LogStash::ConfigurationError, "each host definition must have a \"host\" option") if !host.is_a?(Hash) || host["host"].nil?
    end
  end
  
  def validate_tables!
    if !@tables.nil?
      @tables.each do |table_entry|
        raise(LogStash::ConfigurationError, "each table definition must have a \"name\" option") if !table_entry.is_a?(Hash) || table_entry["name"].nil?
      end
    end
  end

  def validate_strip!
    raise(LogStash::ConfigurationError, "you can not specify both oid_root_skip and oid_path_length") if @oid_root_skip > 0 and @oid_path_length > 0
  end

  ##
  # The StoppableIntervalRunner is capable of running a block of code at a
  # repeating interval, while respecting the stop condition of the plugin.
  class StoppableIntervalRunner
    ##
    # @param plugin [#logger,#stop?]
    def initialize(plugin)
      @plugin = plugin
    end

    ##
    # Runs the provided block repeatedly using the provided interval.
    # After executing the block, the remainder of the interval if any is slept off
    # using an interruptible sleep.
    # If no time remains, a warning is emitted to the logs.
    #
    # @param interval_seconds [Integer,Float]
    # @param desc [String] (default: "operation"): a description to use when logging
    # @yield
    def every(interval_seconds, desc="operation", &block)
      until @plugin.stop?
        start_time = Time.now

        yield

        duration_seconds = Time.now - start_time
        if duration_seconds >= interval_seconds
          @plugin.logger.warn("#{desc} took longer than the configured interval", :interval_seconds => interval_seconds, :duration_seconds => duration_seconds.round(3))
        else
          remaining_interval = interval_seconds - duration_seconds
          sleep(remaining_interval)
        end
      end
    end

    # @api private
    def sleep(duration)
      Stud.stoppable_sleep(duration) { @plugin.stop? }
    end
  end
end
